### DISCUSSION TITLE
How to pass HSE Crystal speed to ArduinoCore / CMSIS

### DISCUSSION THREAD

**Discussion by PerennialNovice** [stm32duino/Arduino_Core_STM32#2322] [Q&A] [answered]
I want to use an external crystal as system clock source, and while I managed to get it set up and implement a SystemClock_Config function that overrides the weak default, `HAL_RCC_GetHCLKFreq()` still returns the wrong frequency.

This is because I use a 16MHz crystal, but somewhere must be the information that a 8MHz crystal is used, I suspect some `STM32halxxyy_default_conf.h` file...

I tried adding `#define HSE_VALUE 16000000U;` to hal_conf_extra.h in my project nad it turns out that this did the trick.

But is it the preferred/best/safe way to do this?

**Comment by fpistm** [comment]
This is specified in the Wiki:
https://github.com/stm32duino/Arduino_Core_STM32/wiki/HAL-configuration#customize-hal-or-variant-definition

**Comment by PerennialNovice** [comment, marked as answer]
Oh, I must have overread that since I checked those pages...

so, about everything that is in the `stm32yyxx_hal_conf_default.h` can be used in `hal_conf_extra.h` as well?

Thank you!
